### ISSUE TITLE
Annotated Self changes variance of class

### ISSUE THREAD

**Issue by Matt-Ord** [astral-sh/ty#2074]
### Summary

```python
class A[M: BasisMetadata = BasisMetadata]:
    def __init__(self, metadata: M) -> None:
        self._metadata = metadata

    def makes_it_covariant(self) -> M:
        """An example method that makes the object covariant."""
        return self._metadata

    def makes_it_invariant(self: A[Any]) -> None:
        """This (incorrectly?) makes the object invariant."""
```

I've use this along with a similar pattern like
```py
    def makes_it_invariant[M1: BasisMetadata = BasisMetadata](
        self: A[M1], other: M1
    ) -> Never:
        """This (incorrectly?) makes the object invariant."""
```

### Version

0.0.2

**Comment by AlexWaygood** [comment]
Here's a [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.13&reportUnreachable=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChRIokBPXLUgc2zwJEAgqXaNGAYwA2xevSgAhWVnoBZNFQAm1YgC4oAOiOSZcqAGUEAI3VadACiX0Vt4tqoBKfUYPjps%2BSEAbVV9JxcNNx0AXV1GKASoTTRgWBhuLCQ4e3o0KWAAGigISPc9KFUPKABaAD4oADk6NDjEtqhc-IMYErsqKABeYtKdcTbk1IhiAGs0enSsiTAAN2IQLDIkHLzgKrqK1vaEgCJTkSg0AA9iPCk0YaQACzBNNkfqYpm5t-uwKwArNASIhLVbrTYGU7HeJHKDoFAgUgdHbdXpRKhjRITT6zeaZdKkMEbUhbTrAfTBETsaJ7epNUgtGFHKEAFUeKig9m4SxA6GBUnYAH4qlNcT8oH9AcDsIS1sSkJDTuJscB7JcKUFwmoRlQad5jIxsTw1RrLDYdcQ9UyoKrLh5GEA) showing that pyright infers this class as covariant, whereas we [do not](https://play.ty.dev/283289c4-acee-4763-8f30-e81c3d650de3). I wondered if pyright's behaviour might be because the `self` parameter is annotated with `self: A[Any]` rather than `self: A[M]`, but pyright has the same behaviour if you change the parameter annotation to `self: A[M]`. It looks like it is applying special handling due to the fact that it is the `self` parameter specifically.

**Comment by carljm** [comment]
I think there are actually two bugs here:

1. It seems like we wrongly consider typevars that are in the generic context of an annotation but specialized away, as "used" in the annotation for purposes of variance inference. If this were fixed, then `A[Any]` (or `A[WhateverConcreteType]`) in any method parameter annotation should not cause invariance of `M`, but `A[M]` should.
2. `self` is different; type checkers ignore it for variance purposes because it is not provided explicitly. (Well, there's some subtlety here around use of unbound methods on `type[...]` types, but `type[...]` types are pretty unsound anyway...) It looks like maybe we don't do that, either, since `self: A[M]` doesn't make `A` invariant in `M` for other type checkers, but does for us.

**Comment by carljm** [comment]
Point 1 is #1459

**Comment by sharifhsn** [comment]
Closed by https://github.com/astral-sh/ruff/pull/25124

Repro:

```py
from typing import Any, Never

class BasisMetadata: ...
class Sub(BasisMetadata): ...

class A[M: BasisMetadata = BasisMetadata]:
    def __init__(self, metadata: M) -> None:
        self._metadata = metadata
    def makes_it_covariant(self) -> M:
        return self._metadata
    def makes_it_invariant(self: "A[Any]") -> None: ...

class B[M: BasisMetadata = BasisMetadata]:
    def __init__(self, metadata: M) -> None:
        self._metadata = metadata
    def makes_it_covariant(self) -> M:
        return self._metadata
    def makes_it_invariant[M1: BasisMetadata = BasisMetadata](
        self: "B[M1]", other: M1
    ) -> Never:
        raise NotImplementedError

a_base: A[BasisMetadata] = A(Sub())  # should be OK if A is covariant
b_base: B[BasisMetadata] = B(Sub())  # should be OK if B is covariant
```

ty 0.0.36:
```
error[invalid-assignment]: Object of type `A[Sub]` is not assignable to `A[BasisMetadata]`
info: `A` is invariant in its type parameter
error[invalid-assignment]: Object of type `B[Sub]` is not assignable to `B[BasisMetadata]`
info: `B` is invariant in its type parameter
Found 2 diagnostics
```

ty 0.0.37:
```
All checks passed!
```
